prusti-specs/src/specifications/preparser.rs

```diff
@@ -289,557 +58,547 @@ impl Parse for SpecEntArgs {
     }
 }
 
-/// The structure to parse Prusti assertions.
-///
-/// Check common::AssertionKind to see all types of Prusti assertions.
-///
-/// Since `syn` can only parse the input as a whole, a preparsing phase
-/// that recognizes the custom Prusti operators and keywords is needed. After
-/// the input is split according to those, it can be parsed using `syn`,
-/// and then is sticked back together to form Prusti assertions.
-/// For a more high-level overview, please check `mod.rs`.
 pub struct Parser {
-    /// The helper to manipulate input.
-    input: ParserStream,
-    /// Members of the conjunction.
-    conjuncts: Vec<AssertionWithoutId>,
-    /// Currently being parsed Rust expression.
-    expr: Vec<TokenTree>,
-    /// A flag to denote whether the previous expression is already resolved
-    /// (parsed into a conjunct).
-    previous_expression_resolved: bool,
-    /// A flag to denote that once the current expression finishes, an operator
-    /// will be expected.
-    expected_operator: bool,
-    /// A flag to denote that the next token must be an operator.
-    expected_only_operator: bool,
-    /// A flag to denote that the parser is currently parsing a pledge
-    /// containing a lhs. This is important so that the parser stops at the
-    /// comma in between lhs and rhs.
-    parsing_pledge_with_lhs: bool,
+    /// Tokens yet to be consumed
+    tokens: VecDeque<TokenTree>,
+    /// Span of the last seen token
+    last_span: Option<Span>,
+    /// Span of the surrounding token
+    source_span: Option<Span>,
 }
 
 impl Parser {
+    /// initializes the parser with a TokenStream
     pub fn from_token_stream(tokens: TokenStream) -> Self {
-        let input = ParserStream::from_token_stream(tokens);
         Self {
-            input,
-            conjuncts: Vec::new(),
-            expr: Vec::new(),
-            previous_expression_resolved: false,
-            expected_operator: false,
-            expected_only_operator: false,
-            parsing_pledge_with_lhs: false,
+            tokens: tokens.into_iter().collect(),
+            last_span: None,
+            source_span: None,
         }
     }
-    fn from_parser_stream(input: ParserStream) -> Self {
+    /// initializes the parser with a TokenStream and the last span
+    fn from_token_stream_and_span(tokens: TokenStream, span: Option<Span>) -> Self {
```

Comment:
All 5 usages of this function give `self.last_span` as the second argument. Unless you expect this to ever not be the case, this function can become `from_token_stream_and_last_span` (or `from_token_stream_last_span`) and take one fewer argument.

Reply:
Addressed by c77b6d0b1638c36b81eddb1e87af37c3f1fff212